### DISCUSSION TITLE
Error: U.S. Census Bureau API Key

### DISCUSSION THREAD

**Discussion by edchanonline** [earthlab-education/earth-analytics-02-python-2024#4] [Q&A01 Urban Greenspace | Cloud Data] [answered]
Has anybody run into the following error within minutes of receiving their API Key from U.S. Census Bureau? 

```You've attempted to validate an unknown key. If it has been more than 48 hours since you submitted your request for this API key then the request has been removed from the system. Please [request a new key](https://api.census.gov/data/key_signup.html) and activate it within 48 hours.```

![Screenshot 2024-01-23 at 2 50 20 PM](https://github.com/earthlab-education/earth-analytics-02-python-2024/assets/4795392/e78c673d-ac35-4384-838e-aa640f13cfae)

**Comment by misterskye** [comment, marked as answer]
I got the same error, but then I tried the key a second time, and it worked.  Maybe I just got lucky, but I would definitely try.

**Reply by edchanonline** [reply]
It was so odd. I didn't receive an activation confirmation of the key, but it appeared to be working! Thanks Skye!!

**Comment by advythr** [comment]
Yes, I had the same problem but the activation link worked after a few hours.

**Comment by eculler** [comment]
I also had this problem. I ended up trying again and it worked…I was hoping it wouldn’t be a consistent issue. I had to sign up again with a different email (I used my Identikey one instead of my usual email), and then it worked. But it seems like folks are running into this for all sorts of emails.

On this subject — You should avoid putting your API keys up on GitHub. We will discuss on Thursday how to avoid committing your API key using the getpass library. Not a huge deal in this case, but you don’t want a bot wrecking your account.

**Reply by edchanonline** [reply]
Good call out. I already wrote a function to tackle the perspective of saving API keys locally.

```
def get_api_key(service_name, username, override=False):
    # Get the API key from the keyring
    api_key = keyring.get_password(service_name, username)

    # If the API key is None or override is True, prompt for a new key
    if api_key is None or override:
        api_key = getpass.getpass(prompt='Enter U.S. Census Bureau API Key: ')
        keyring.set_password(service_name, username, api_key)

    return api_key
```


    
How to use it:

```
service_name = "US_Census_Bureau_ACS"
username = "api_key"

# API key expires in 48 hours. Set override to True to force a new key.
api_key = get_api_key(service_name, username, override=False)
```

**Reply by edchanonline** [reply]
And, import these:

```
import getpass
import keyring
```

**Reply by eculler** [reply]
Cool! Yeah, that's basically what I do in earthpy. I can't get keyring to work on Codespaces, though, so I put in an option to save to a file in the home folder.
